runtime/src/snapshot_utils.rs

```diff
@@ -819,7 +813,11 @@ pub fn get_bank_snapshots(bank_snapshots_dir: impl AsRef<Path>) -> Vec<BankSnaps
                         bank_snapshots.push(snapshot_info);
                     }
                     Err(err) => {
-                        error!("Unable to read bank snapshot for slot {}: {}", slot, err);
+                        error!(
+                            "Unable to read bank snapshot for slot {}: {}.  Removing the dir",
+                            slot, err
+                        );
+                        remove_bank_snapshot(slot, &bank_snapshots_dir);
```

Comment:
get_bank_snapshots() does the work of scanning the directories under the top bank_snapshots_dir and return the vector of snapshots (BankSnapshotInfo) for the snapshot dirs not returning any error.  This is called in purge_old_bank_snapshots() and do_get_highest_bank_snapshot()

The previous behavior is that for any snapshot dirs which returns error, it generates an error! log and leaves that bad dir on disk.

This new behavior is to always remove any snapshot dir whenever reading from it returns an error, failing to return a valid BankSnapshotInfo.

I think this is the clean way, ensuring there is no any leftover bad dir on the disk.

Reply:
I think your concern is that the function name get_bank_snapshots() is misleading, doing cleaning work in additional to getting.  I agree on that.  The real clean way could be as below
Have a SnapshotTopDir struct.  
It has the member functions 
fn scan_snapshots_dir(): does scanning and cleaning, and building a vector of BankSnapshotInfos.
fn get_snapshots(): return the built vector of BankSnapshotInfos.

Reply:
Actually in one case the software process could fail without this.

Suppose the account paths are in the memory mounted file system, taking up the memory space.  The software process is killed when add_bank_snpashot() is almost done, but not done yet to have the completion file flag set up, so it leaves an incomplete snapshot dir on disk with its account fles taking much memory space.  Then, without this cleanup step, loading the new bank could fail due to Out of Memory (OOM).

Reply:
Back in Add remove_incomplete_bank_snapshot_dir (#31131), the handling on incomplete snapshot dir was added.  This is an improved, more solid version of that.

Reply:
In PR "Add remove_incomplete_bank_snapshot_dir" (https://github.com/solana-labs/solana/pull/31131), the first commit [aaafef7](https://github.com/solana-labs/solana/pull/31131/commits/aaafef74803ca4e2050b8cd83ef352c5a59509b4) was to have a separate function to to purge invalid snapshot dirs.  
That scans through the snapshots dir and delete the invalid ones.  

I reverted that change, and opted to delete the dir in-place inside get_bank_snapshots() because that directory scanning part is duplicated logic, same to the scanning part in get_bank_snapshots.  I think it is much simpler just to scan a directory in one place, looking at the scanning result for once, either accept it and construct a BankSnapshotInfo for it, or reject it and cleanup its disk space.

PR 31131 still leaves two issues.
1.  The completion check is not done as the 1st check.  An incomplete dir caused some other IO failures before reaching the completion check, so the directory was not removed.  It is later fixed by "Check incomplete file flag first #31409"
2. The handling of incompletion is to call fs::remove_dir_all(bank_snapshot_dir) to remove the dir, leaving the accounts files orphaned.  The clean_orphaned_account_snapshot_dirs call at the next process start time will clean up them.  It is fine when they are using the disk space.  But in the case memory-mounted account-paths, this can cause OOM.  
The reason of calling remove_dir_all instead of remove_bank_snapshot is that I was not sure if it is good for a snapshot dir in bad state.  I am afraid it may just raise some IO errors and quit, leaving the dir unremoved.  This PR is to address that concern, make sure it can remove the dir regardless of the state of the dir.  Then I can call it on any bad dir and make sure the dir is cleanly removed.